pkg/api/fulcioca_signing_cert.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package api
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/go-openapi/runtime/middleware"
+	"github.com/spf13/viper"
+
+	"github.com/sigstore/fulcio/pkg/ca/fulcioca"
+	"github.com/sigstore/fulcio/pkg/ctl"
+	"github.com/sigstore/fulcio/pkg/generated/restapi/operations"
+	"github.com/sigstore/fulcio/pkg/log"
+	"github.com/sigstore/fulcio/pkg/oauthflow"
+	"github.com/sigstore/fulcio/pkg/pkcs11"
+)
+
+func FulcioCASigningCertHandler(params operations.SigningCertParams, principal *oidc.IDToken) middleware.Responder {
+
+	// none of the following cases should happen if the authentication path is working correctly; checking to be defensive
+	if principal == nil {
+		return handleFulcioAPIError(params, http.StatusInternalServerError, errors.New("no principal supplied to request"), invalidCredentials)
+	}
+
+	emailAddress, emailVerified, err := oauthflow.EmailFromIDToken(principal)
+	if !emailVerified {
+		return handleFulcioAPIError(params, http.StatusBadRequest, errors.New("email_verified claim was false"), emailNotVerified)
+	} else if err != nil {
+		return handleFulcioAPIError(params, http.StatusBadRequest, err, invalidCredentials)
+	}
+
+	publicKey := *params.CertificateRequest.PublicKey
+	pkixPubKey, err := x509.ParsePKIXPublicKey(*publicKey.Content)
+	if err != nil {
+		return handleFulcioAPIError(params, http.StatusBadRequest, err, malformedPublicKey)
+	}
+
+	// Perform a proof challenge verification
+	if err := fulcioca.CheckSignature(pkixPubKey, *params.CertificateRequest.SignedEmailAddress, emailAddress); err != nil {
+		return handleFulcioAPIError(params, http.StatusBadRequest, err, invalidSignature)
+	}
+
+	// Use a generic Cert failure message as we don't want to tell the client about a HSM failure and risk exposure of
+	// internal security controls
+	p11Ctx, err := pkcs11.InitHSMCtx()
+	if err != nil {
+		return handleFulcioAPIError(params, http.StatusInternalServerError, err, failedToCreateCert)
```

Comment:
change `err` to `errors.Wrap(err, "initializing HSM context")` so when this gets logged you have some additional information present

Reply:
done